File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def isSameTree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
            if not p and q:
                return False
            elif not q and p:
                return False
            elif not q and not p:
                return True
            else:
                return p.val == q.val and isSameTree(p.left, q.left) and isSameTree(p.right, q.right)

        queue = deque([root])
        while queue:
            node = queue.popleft()
            if isSameTree(node, subRoot):
                return True
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        return False
            
        

            

File: Data Structures & Algorithms/path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

    bool dfs(TreeNode* node, int curr, int target){
        if (node){
            curr += node->val;
            if ((curr == target) && (node->left == nullptr && node->right == nullptr)){
                return true;
            }
            else{
                return dfs(node->left, curr, target) || dfs(node->right, curr, target);
            }
        }
        return false;
    }

public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, 0, targetSum);
    }
};